### DISCUSSION TITLE
How to get user from jwt token call? (WebSocket)

### DISCUSSION THREAD

**Discussion by ervnjmsdnts** [fastapi-users/fastapi-users#808] [Q&A] [answered]
Hello I am currently trying to integrate WebSocket with this package. The method of authentication that I am using is JWT so I need to get the token from the authorization header. I tried to get it with this:

```
@app.websocket("/ws")
async def websocketEndpoint(websocket: WebSocket, authorization: str = Header(...)):
    schema, param = get_authorization_scheme_param(authorization)

    user = await jwt_authentication(param, user_manager=getUserManager())

    await websocket.accept()

    while True:
        data = await websocket.receive_text()
        print(f"Data: {data}")
```

But it is throwing me this error:
```
AttributeError: 'async_generator' object has no attribute 'get'
```

If you are wondering where I got this code. I got it from this discussion of this package:
https://github.com/fastapi-users/fastapi-users/discussions/436

**Comment by frankie567** [comment, marked as answer]
The problem here is that you are calling `getUserManager` directly. This dependency doesn't return a `UserManager` instance directly: it's a generator. Inside FastAPI, you don't notice it because FastAPI is smart enough to unpack it inside its dependency injection system.

If you want to call it directly, I suggest you to read the [Create a user programmatically cookbook](https://fastapi-users.github.io/fastapi-users/cookbook/create-user-programmatically/): in the part 1, you'll find some insights and guidance on how to get your UserManager manually.

Basically, it'll look like this:

```py
get_user_db_context = contextlib.asynccontextmanager(get_user_db)
get_user_manager_context = contextlib.asynccontextmanager(get_user_manager)

@app.websocket("/ws")
async def websocketEndpoint(websocket: WebSocket, authorization: str = Header(...)):
    schema, param = get_authorization_scheme_param(authorization)

    async with get_user_db_context() as user_db:
        async with get_user_manager_context(user_db) as user_manager:
            try:
                user = await jwt_authentication(param, user_manager=user_manager)
            except:
                # Invalid session / not existing user
                await websocket.close(code=1000)
                return

    await websocket.accept()

    while True:
        data = await websocket.receive_text()
        print(f"Data: {data}")
```

**Reply by ervnjmsdnts** [reply]
Cool thanks! But I did try this scuff solution, I don't know if it is optimally correct though.
```
@app.websocket("/ws")
async def websocketEndpoint(websocket: WebSocket, authorization: str = Header(...)):
    schema, param = get_authorization_scheme_param(authorization)

    payload = jwt.decode_jwt(param, TOKEN_SECRET, audience=["fastapi-users:auth"])

    user = await user_collection.find_one({"id": UUID(payload["user_id"])})

    if user:
        await websocket.accept()

        while True:
            data = await websocket.receive_text()
            print(f"Data: {data}")
```
As you can see I tried to just decode the token and get the id of the user and then find it in the database

**Reply by frankie567** [reply]
Yes, that's another possibility; this is basically what happens under the hood of the authentication class.
